hedgehog/test/internals.hs

```diff
@@ -0,0 +1,87 @@
+module Main where
+
+import Control.Applicative
+import Control.Monad.Trans.Maybe
+import Data.Functor.Classes
+import Hedgehog.Internal.Gen
+import Hedgehog.Internal.Range
+import Hedgehog.Internal.Seed
+import Hedgehog.Internal.Tree
+import Test.QuickCheck.Classes
+import Test.QuickCheck.Checkers
+import Test.QuickCheck
+import Test.Tasty
+import Test.Tasty.ExpectedFailure
+import Test.Tasty.QuickCheck
```

Comment:
I was just thinking that it *may* be possible to let Hedgehog itself run those tests instead of taking on tasty and friends. I like the simplicity of not having to take on any external dependencies:

```haskell
{-# LANGUAGE OverloadedStrings #-}
module Main (
    main
  ) where

import           Control.Monad  (unless)
import           System.Exit    (exitFailure)

import           Hedgehog
import qualified Hedgehog.Gen   as Gen
import qualified Hedgehog.Range as Range

main :: IO ()
main = do
  results <- sequence [
      properties
    , regressionTests
    ]

  unless (and results)
    exitFailure

properties :: IO Bool
properties =
  checkParallel $ Group "Properties" [
        ("Property A"
        , property success)
      , ("Property B"
        , property discard)
      , ("Property C"
        , property failure)
    ]

regressionTests :: IO Bool
regressionTests =
  checkParallel $ Group "Regression" $
    do
    (a, b) <-
      [
        (1, 1),
        (2, 2),
        (3, 3)
      ]
    return
      ("Regression A"
      , withTests 1 . property $ a === b)
    ++

    do
    (a, b) <-
      [
        (1, 1),
        (2, 2),
        (3, 0) -- Uh-oh!
      ]
    return
      ("Regression B"
      , withTests 1 . property $ a === b)
```

It surely isn't perfect (not sure we can *recheck* a failing test), but I thought it'd be nice  to share.

---

It also feels kind of weird having to use QC to test Hedgehog internals!

Reply:
(This is just an idea, and shouldn't delay further the PR.)

Reply:
> I was just thinking that it may be possible to let Hedgehog itself run those tests instead of taking on tasty and friends. I like the simplicity of not having to take on any external dependencies:

Hmm, my intuition would be you shouldn't base tests for a test framework on the very same test framework. If the test framework has a bug, the internal tests might report false positives.